spec/src/main/asciidoc/servlet-spec-body.adoc

```diff
@@ -1300,6 +1300,63 @@ the header value to an `int`, a `NumberFormatException` is thrown. If
+HTTP/0.9:: The URI is the `document address` as defined by the W3C [Original HTTP](https://www.w3.org/Protocols/HTTP/AsImplemented.html) document.
+
+HTTP/1.0:: The URI is extracted from the `Request-URI` in the `Request-Line` as defined by [RFC 1945](https://datatracker.ietf.org/doc/html/rfc1945#section-5.1). URIs in `abs_path` form are passed unchanged to subsequent steps. URIs in `absoluteURI` have the protocol and authority removed to convert them to `origin-form` and are then passed the subsequent steps.
+
+HTTP/1.1:: The URI is extracted from the `request-target` as defined by [RFC 7230](https://datatracker.ietf.org/doc/html/rfc7230#section-3.1.1). URIs in `origin-form` are passed unchanged to subsequent steps. URIs in `absolute-form` have the protocol and authority removed to convert them to `origin-form` and are then passed the subsequent steps. URIs in `authority-form` or `asterisk-form` are outside of the scope of this specification.
+
+HTTP/2:: The URI is the `:path` pseudo header as defined by [RFC 7540](https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.3) and is passed unchanged to stage 2.
+
+HTTP/3:: The URI is the `:path` pseudo header as currently defined by the [draft RFC](https://datatracker.ietf.org/doc/html/draft-ietf-quic-http-34).
+
+Other protocols:: Containers may support other protocols. Containers should extract an appropriate URI for the request from the protocol and pass it to stage 2.
+
+==== Separation of path and query
+The URI is split by the first occurrence of any '?' character to path and query.  The query is preserved for later handling and the following steps applied to the path.
+
+==== Discard fragment
+A fragment in the path is indicated by the first occurrence of a `\#` character.  Any `#` character and following fragment is removed from the path and discarded.
+
+==== Decoding of characters as UTF-8
+Characters other than `/`, `;` and `%` that are encoded in `%nn` form are decoded. The resulting octet sequences is treated as UTF-8 and converted to a character sequence.
+
+==== Collapse sequences of multiple `"/"` characters
+Any sequence of more than one `"/"` character in the URI must be replaced with a single `"/"`.
+
+==== Remove dot-segments+
+* Sequences of the form `"/./"` must be replaced with `"/"`.
+* Sequences of the form `"/" + segment + "/../"` must be replaced with `"/"`.
+
+==== Removal of path parameters
+A path segment containing the `";"` character is split at the first occurence of `";"`. The segment is replaced by the character sequence preceeding the `";"`. The characters following the `";"` are considered a path parameters and may be preserved by the container for later processing (eg `jsessionid`).
+
+==== Decoding of remaining `%nn` sequences
+Any remaining `%nn` sequences in the path should be decoded. Some containers may be configured to leave some specific characters encoded (eg. the characters '/' and '%' may be left decoded by some container configuration).
+
```

Comment:
I think that rather than describing an algorithm that has the effect that we want regarding segments, we should be more direct and describe an algorithm in terms of segments.    If we ultimately make a new segment based API available, then it can be in terms of the segments described here and we don't need to re-specify.

```suggestion
==== Split path into segments
The path is split into multiple segments divided by the `/` character.

==== Remove path parameters
Any segment containing the `";"` character is split at the first occurrence of `";"`. The segment is replaced by the character sequence preceding the `";"`. The characters following the `";"` are considered a path parameters and may be preserved by the container for later decoding and/or processing (eg `jsessionid`).

==== Decode characters
Characters that are encoded in `%nn` form are decoded in each segment. The resulting octet sequences is treated as UTF-8 and converted to a character sequence.

==== Remove dot-segments
All segments that are exactly `"."` are removed from the segment series. Segments that are exactly '".."` and that are preceded by a non `".."` segment are removed together with the preceding segment.     This canonicalization differs from RFC3986 in that segments with `%` encoded dots and/or parameters may be treated as dot segments.

==== Remove Empty Segments
By default, empty segments are removed.  Containers may be configured to retain empty segments.

==== Concatenate segments
The segments are concatenated into a single path string with each segment preceded by the `"/"` character.  If a segment contains the `"/"` or `"%"` characters, and the request will not be rejected below, then the container may re-encode those characters to `%nn` form.
```

Param removal is before the decoding step, which is before the dot removal step.  This is the main deviation from the RFC, so I have noted it in the text.

Note that it is immaterial if an implementation actually decodes then re-encodes `%2F` or just leaves it encoded.  I think it reads better this way, as it makes leaving `%2F` in the exception case.

I have switched dot segment removal before empty segment removal. My reason is that if a container wishes to retain empty segments then the result is more consistent with RFC3986.

Note this algorithm "forgets" if a path did not start with '/'.  I don't think that is a problem as we can probably specify that it must start with a '/' in step 1, and then add a rejection below as defence in depth.

Reply:
I'm going to merge this suggestion to the PR now, because I think we have resolved most of the behaviour differences and it is now just a matter of how we describe it in text.  I'd like to do some cleanup of this and it is getting a bit meta editing a suggestion of a PR....   
Merging doesn't mean I think this text is settled and still happy to discuss and/or revert to previous form if that is the consensus.

Reply:
@stuartwdouglas The third option when `/%2e%2e/` is allowed is to throw if `getPath()` or `getPathInfo()` are called.   That way it is only allowed if `getRequestUri()` or new API is used.  Actually `getRequestUri()` can still be dangerous with this... so perhaps if it is allowed it is only for new API?

If that is the case, we can move decoding until after dot-segment handling.

Note as this comment is outdated. I'm going to resolve and then re-comment on latest text.